File: crypto/generateUrlPassword.rb
#!/usr/bin/ruby

# Written by Bill Brassfield in 2014

#
# A technically non-random but interesting and fun way to generate
# secure passwords for your online accounts.  (It's "secure" as long
# as your Master Password is secure.  For added security, the Serial
# Number doesn't have to be numeric -- it can be any string, so feel
# free to use your own secret serial-numbering scheme to come up with
# new/updated passwords.)  Also, you might want to append your account's
# username and/or e-mail address to the end of the URL (or put it at
# the beginning before the URL).  Whatever you come up with, just be
# consistent so you won't forget how to regenerate your password if
# you forget it.
#

require 'digest'
require 'base64'

print "URL: "
url = STDIN.gets.chomp

print "Master Password: "
system "stty -echo"
masterPass = STDIN.gets.chomp
system "stty echo"
puts

print "Serial #: "
serial = STDIN.gets.chomp

print "Use Special Characters? "
useSpecialChars = STDIN.gets.chomp

print "Password Length (up to 24 chars) [16]: "
passLen = STDIN.gets.chomp
if (passLen =~ /^\s*$/)
  passLen = 16
else
  passLen = Integer(passLen)
end

regularChars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789/+'
specialChars = <<'SPECIAL_CHARS'
`~!@#$%^&*()-_=?[{]}\|;:'",<.>/+
SPECIAL_CHARS
specialChars = specialChars.chomp
regToSymMap = Hash.new()
regularArr = regularChars.split(//)
specialArr = specialChars.split(//)
0.upto(63) { |i|
  regToSymMap[regularArr[i]] = specialArr[i % 32]
}
regToSymMap['='] = '='

sha256 = Digest::SHA256.new

firstPacked  = sha256.digest(url + " " + masterPass + " " + serial + "\n")
# firstHash    = firstPacked.unpack("H*").first
firstHash    = Digest::hexencode(firstPacked)
secondPacked = sha256.digest(firstHash + " " + masterPass + " " + serial + "\n")
 
encoded = Base64.encode64(firstPacked)
encoded = encoded.gsub(/\n/, "")

posMap   = Hash.new()
posCount = 0

secondPacked.split(//).each do |ch|
  pos = ch.ord
  if (pos > 251)
  else
    pos = pos % 42
    if (defined?(posMap[pos]) && posMap[pos] == 1)
    else
      posMap[pos] = 1
      posCount += 1
    end
    if (posCount >= 14)
      break
    end
  end
end

encArr = Array.new()

if (useSpecialChars =~ /^\s*(Y|y|1)\s*$/)
  encArr = encoded.split(//)
  0.upto(encArr.length - 1) { |i|
    if (defined?(posMap[i]) && (posMap[i] == 1))
      encArr[i] = regToSymMap[encArr[i]]
    end
  }
else
  encoded = encoded.gsub(/[^A-Za-z0-9]/, '')
  encArr = encoded.split(//)
end

password = encArr[0 .. (passLen-1)].join('')

puts
puts password
puts

File: crypto/generateUrlPassword.pl
#!/usr/bin/perl -w

# Written by Bill Brassfield in 2014

#
# A technically non-random but interesting and fun way to generate
# secure passwords for your online accounts.  (It's "secure" as long
# as your Master Password is secure.  For added security, the Serial
# Number doesn't have to be numeric -- it can be any string, so feel
# free to use your own secret serial-numbering scheme to come up with
# new/updated passwords.)  Also, you might want to append your account's
# username and/or e-mail address to the end of the URL (or put it at
# the beginning before the URL).  Whatever you come up with, just be
# consistent so you won't forget how to regenerate your password if
# you forget it.
#

use strict;

use Digest::SHA qw(sha256 sha256_hex sha256_base64);
use MIME::Base64;

$|++;

my $defaultLength = 16;
my $maxPassLength = 24;

my $regularChars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789/+';

my $specialChars = <<'SPECIAL_CHARS';
`~!@#$%^&*()-_=?[{]}\|;:'",<.>/+
SPECIAL_CHARS
chomp $specialChars;

my @regularArr = split('', $regularChars);
my @symbolArr  = split('', $specialChars);

my %regToSymMap = ();
for (my $i=0; $i<64; $i++)
{
  $regToSymMap{$regularArr[$i]} = $symbolArr[$i % 32];
}
$regToSymMap{'='} = '=';

print "URL: ";
my $url = <STDIN>;
chomp $url;

system "stty -echo";
print "Master Password: ";
my $masterPass = <STDIN>;
chomp $masterPass;
system "stty echo";
print "\n";

print "Serial #: ";
my $serial = <STDIN>;
chomp $serial;

print "Use Special Characters? ";
my $useSpecialChars = <STDIN>;
chomp $useSpecialChars;

my $passwordLength = -1;

while (($passwordLength < 4) || ($passwordLength > $maxPassLength))
{
  print "Password Length (up to $maxPassLength chars) [$defaultLength]: ";
  $passwordLength = <STDIN>;
  chomp $passwordLength;

  if ($passwordLength =~ /^\s*$/s)
  {
    $passwordLength = $defaultLength;
  }
}

my $firstPacked  = sha256("$url $masterPass $serial\n");
my $firstHash    = unpack("H*", $firstPacked);
my $secondPacked = sha256("$firstHash $masterPass $serial\n");

my $encoded = encode_base64($firstPacked);
$encoded =~ s/\n//g;

print "\n";

my %posMap   = ();
my $posCount = 0;
foreach my $ch (split('', $secondPacked))
{
  my $pos = ord($ch);
  next if ($pos > 251);
  $pos = $pos % 42;
  if (!exists($posMap{$pos}))
  {
    $posMap{$pos} = 1;
    $posCount++;
  }
  last if ($posCount >= 14);
}

my @encArr;

if ($useSpecialChars =~ /^\s*(?:Y|1)\s*$/i)
{
  @encArr = split('', $encoded);

  for (my $i=0; $i<=$#encArr; $i++)
  {
    if (exists($posMap{$i}))
    {
      $encArr[$i] = $regToSymMap{$encArr[$i]};
    }
  }
}
else
{
  $encoded =~ s/[^A-Za-z0-9]//g;
  @encArr = split('', $encoded);
}

my $password = join('', @encArr[0 .. ($passwordLength-1)]);

print "$password\n\n";

